### PR TITLE
Replace *-whitelist/*-blacklist/*-requirelist rules with *-allowed-list/*-disallowed-list/*-required-list rules

### LINKED ISSUE #4844: Add new names and alias *-*list rules
> What is the problem you're trying to solve?

**Short:** Provide terminology to package users that does not perpetuate white supremacy and racism against Black people

**Long:** There has been a much-needed shift in the tech industry to transition our naming conventions away from problematic racist terminology. For example, [Github is replacing the convention of having a master branch to remove any master/slave reference](https://www.cnet.com/news/microsofts-github-is-removing-coding-terms-like-master-and-slave/). 

Other terminology that has been pointed out as problematic are the terms blacklist and whitelist. They tell us that black is bad and white is good. Some of our teams at IBM have starting switch to terms like [blocklist](https://cloud.ibm.com/docs/overview?topic=overview-glossary#x3954055) and [allowlist](https://cloud.ibm.com/docs/overview?topic=overview-glossary#x3954001).

This much-appreciated and foundational tool uses blacklist and whitelist terms significantly in its rule names and it would be good to change them.

> What solution would you like to see?

Replace blacklist with blocklist and whitelist with allowlist in the rule names maintained in this repo. Additionally, provide supportive deprecation warnings so developers know about the change and why it is an important change.

### PR DESCRIPTION
**Pull request by kevindew**
> Which issue, if any, is this issue related to?

Resolves #4844

> Is there anything in the PR that needs further explanation?

This PR serves as a potential implementation route to achieve the re-naming change raised in issue #4844, where I expect there is still to be a reasonable amount of discussion on the issue. It has been reworked following @jeddy3 [suggestion](https://github.com/stylelint/stylelint/issues/4844#issuecomment-650574763) on an implementation route.

It takes an approach of renaming all `*-blacklist`, `*-requirelist` and `*-whitelist` rules to `*-disallowed-list`, `*-required-list` and `*-allowed-list` which creates a consistent `*-list` pattern for naming list based rules. There have then been aliases created so that the previously named rules remain operational.

### CODE REVIEW

**Comment by jeddy3** [comment]
@kevindew Thank you for getting the ball rolling with this.

> The approach I took for this was to rename the existing rules first (and associated documentation), so that the git history on the rules could be transferred to the new rules

Excellent. Regardless of the final route we decide on, we'll need to go about it this way.

> This PR serves as a potential implementation route to achieve the re-naming change raised in issue #4844, where I expect there is still to be a reasonable amount of discussion on the issue.

I've [made some suggestions](https://github.com/stylelint/stylelint/issues/4844#issuecomment-650574763) about the implementation route and renaming over in the issue.

**Comment by kevindew** [comment]
> @kevindew Thank you for getting the ball rolling with this.
> 
> > The approach I took for this was to rename the existing rules first (and associated documentation), so that the git history on the rules could be transferred to the new rules
> 
> Excellent. Regardless of the final route we decide on, we'll need to go about it this way.
> 
> > This PR serves as a potential implementation route to achieve the re-naming change raised in issue #4844, where I expect there is still to be a reasonable amount of discussion on the issue.
> 
> I've [made some suggestions](https://github.com/stylelint/stylelint/issues/4844#issuecomment-650574763) about the implementation route and renaming over in the issue.

Thanks, the clarity in your suggestions was super helpful. I've now amended this in line with them, do let me know if I've misinterpreted any.

I wasn't too sure of what testing approach, if any, to take with the aliased rules and put [this together](https://github.com/stylelint/stylelint/blob/cbdf5de0b3b61ce73c673266baec9360d399e984/lib/__tests__/aliasedRules.test.js) where I might benefit from some pointers to precedence if it's not sufficiently consistent with this codebase. 

I also wasn't sure if this risks being a BC break given that violations for a whitelist rule will be returned under a new rule name. For example, a unit-whitelist violation will be raised as a unit-allowed-list violation. I expect you'll know better than me if that is considered part of the API.

File: scripts/visual-config.json [hunk 1]
```diff
@@ -1,7 +1,7 @@
 {
 	"rules": {
 		"color-no-invalid-hex": true,
-		"declaration-property-unit-blacklist": { "margin": ["px"] },
+		"declaration-property-unit-disallowed-list": { "margin": ["px"] },
```
Comment (jeddy3) [inline scripts/visual-config.json:4]:
```suggestion
		"unit-disallowed-list": ["px"],
```

Let's:

1. change this to the shorter `unit-disallowed-list`
2. run `cd scripts && ./visual.sh`
3. screenshot the "Normal" terminal output
4. [optimise](https://kraken.io/web-interface) the resultant `.png`
5. replace https://github.com/stylelint/stylelint/blob/master/example.png with it

It's shown in the README and project homepage.

File: docs/user-guide/rules/list.md [hunk 2]
```diff
@@ -101,11 +101,11 @@ Grouped first by the following categories and then by the [_thing_](http://apps.
...
 ### Function
 
-- [`function-blacklist`](../../../lib/rules/function-blacklist/README.md): Specify a blacklist of disallowed functions.
+- [`function-allowed-list`](../../../lib/rules/function-allowed-list/README.md): Specify a list of allowed functions.
+- [`function-disallowed-list`](../../../lib/rules/function-disallowed-list/README.md): Specify a list of disallowed functions.
 - [`function-url-no-scheme-relative`](../../../lib/rules/function-url-no-scheme-relative/README.md): Disallow scheme-relative urls.
-- [`function-url-scheme-blacklist`](../../../lib/rules/function-url-scheme-blacklist/README.md): Specify a blacklist of disallowed URL schemes.
-- [`function-url-scheme-whitelist`](../../../lib/rules/function-url-scheme-whitelist/README.md): Specify a whitelist of allowed URL schemes.
-- [`function-whitelist`](../../../lib/rules/function-whitelist/README.md): Specify a whitelist of allowed functions.
+- [`function-url-scheme-disallowed-list`](../../../lib/rules/function-url-scheme-disallowed-list/README.md): Specify a list of disallowed URL schemes.
```
[the remark this answers is not in the document]
Comment (kevindew) [reply docs/user-guide/rules/list.md]: Whoops, nice spot 👍

File: lib/__tests__/integration.test.js [hunk 3]
```diff
@@ -141,6 +143,19 @@ it('Scss integration test', () => {
 		});
 });
 
+it('Rule aliasing integration test', () => {
```
[the remark this answers is not in the document]
Comment (kevindew) [reply lib/__tests__/integration.test.js]: Sure, done

File: lib/rules/declaration-property-unit-disallowed-list/index.js [hunk 4]
```diff
@@ -12,16 +12,16 @@ const ruleMessages = require('../../utils/ruleMessages');
 const validateOptions = require('../../utils/validateOptions');
```
[the remark this answers is not in the document]
Comment (kevindew) [reply lib/rules/declaration-property-unit-disallowed-list/index.js:12]: This just seems to be GitHub becoming confused when merging the diffs of multiple commits. In the individual commits this rule doesn't shift.For example:

```
rename lib/rules/{declaration-property-unit-blacklist => declaration-property-unit-disallowed-list}/README.md (91%)
rename lib/rules/{declaration-property-unit-blacklist => declaration-property-unit-disallowed-list}/__tests__/index.js (100%)
rename lib/rules/{declaration-property-unit-blacklist => declaration-property-unit-disallowed-list}/index.js (85%)
 ```

 and 

 ```
rename lib/rules/{declaration-property-unit-whitelist => declaration-property-unit-allowed-list}/README.md (92%)
rename lib/rules/{declaration-property-unit-whitelist => declaration-property-unit-allowed-list}/__tests__/index.js (100%)
rename lib/rules/{declaration-property-unit-whitelist => declaration-property-unit-allowed-list}/index.js (85%)
```

File: lib/rules/selector-pseudo-element-allowed-list/README.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
-# selector-pseudo-element-blacklist
```
[the remark this answers is not in the document]
Comment (kevindew) [reply lib/rules/selector-pseudo-element-allowed-list/README.md:1]: Same GitHub session, looks good in commit history.

File: scripts/visual-config.json (hunk 1 above)
Comment (kevindew) [reply scripts/visual-config.json]: Cool, I've added this in https://github.com/stylelint/stylelint/pull/4845/commits/8e8aa2cb3d6403bb802f6169b5242009314eec41

**Comment by kevindew** [comment]
> I think we're good. The only place outside of a config where the rule names are used is within the [`checkAgainstRule` plugin utility](https://stylelint.io/developer-guide/plugins#stylelintutilscheckagainstrule). It would be great if we could [add a test](https://github.com/stylelint/stylelint/blob/master/lib/utils/__tests__/checkAgainstRule.test.js#L24-L42) ensuring the renamed rule is in the warning when an alias is used.

I've added that in https://github.com/stylelint/stylelint/pull/4845/commits/96acce6f48016fd2408f4ffd9581346a1b8376f0 with: https://github.com/stylelint/stylelint/blob/96acce6f48016fd2408f4ffd9581346a1b8376f0/lib/utils/__tests__/checkAgainstRule.test.js#L64-L82

Let me know if you'd like any amends to it (I don't have much of an understanding of the code this part is testing)

**Comment by kevindew** [comment]
Great stuff, yup this needs at least a changelog entry for this repo. Would that be where the write-up goes or somewhere else? 

I’ve not opened any other PRs in other repos but I’ll look into sorting that out now this is on the way in.

**Comment by malsf21** [comment]
> Great stuff, yup this needs at least a changelog entry for this repo. Would that be where the write-up goes or somewhere else?

At least in the `CHANGELOG.md` for sure, though I'd also imagine that something on the website or more user-facing would also be needed? I'm still pretty new around here, so I'd defer to @jeddy3.

> I’ve not opened any other PRs in other repos but I’ll look into sorting that out now this is on the way in.

Sounds good! I can help you with some of the low-hanging but tedious fruit like the config updating!

**Comment by jeddy3** [comment]
> Also, have we made the PR to the stylelint.io website to add redirects for the renamed rules?

Not yet. It's the only thing remaining to do.

> I can help you with some of the low-hanging but tedious fruit like the config updating!

@malsf21 Excellent. If you've time, would you mind rustling up a pull request for the redirects? I've created an issue for it: https://github.com/stylelint/stylelint.io/issues/158

> Great stuff, yup this needs at least a changelog entry for this repo. 

We'll [add a changelog entry just after merging](https://stylelint.io/maintainer-guide/pull-requests#merging). We'll [tweet the release too](https://stylelint.io/maintainer-guide/releases#performing-releases).

**Comment by jeddy3** [comment]
Changelog:

> - Added: `*-allowed-list`, `*-disallowed-list` and `*-required-list` new names for `*-whitelist`, `*-blacklist` and `*-requirelist` rules, respectively; the rules are aliased as their old names ([#4845](https://github.com/stylelint/stylelint/pull/4845)).

**Pull request merged**